plugin/pulpcore/plugin/stages/content_unit_stages.py

```diff
@@ -0,0 +1,176 @@
+
+    Each :class:`~pulpcore.plugin.stages.DeclarativeContent` is sent to after it has been handled.
+
+    This stage drains all available items from `in_q` and batches everything into one large call to
+    the db for efficiency.
+
+    Args:
+        in_q (:class:`asyncio.Queue`): Each item is a
+            :class:`~pulpcore.plugin.stages.DeclarativeContent`
+        out_q (:class:`asyncio.Queue`): Each item is a
+            :class:`~pulpcore.plugin.stages.DeclarativeContent`
+
+    Returns:
+        The content_unit_saver stage as a coroutine to be included in a pipeline.
+    """
+    declarative_content_list = []
+    shutdown = False
+    while True:
+        try:
+            declarative_content = in_q.get_nowait()
+        except asyncio.QueueEmpty:
+            if not declarative_content_list and not shutdown:
+                declarative_content = await in_q.get()
+                declarative_content_list.append(declarative_content)
+                continue
+        else:
+            declarative_content_list.append(declarative_content)
+            continue
+
+        content_artifact_bulk = []
+        remote_artifact_bulk = []
+        remote_artifact_map = {}
+
+        with transaction.atomic():
+            for declarative_content in declarative_content_list:
+                if declarative_content is None:
+                    shutdown = True
+                    continue
+                if declarative_content.content.pk is None:
+                        declarative_content.content.save()
+                        for declarative_artifact in declarative_content.d_artifacts:
+                            content_artifact = ContentArtifact(
+                                content=declarative_content.content,
+                                artifact=declarative_artifact.artifact,
+                                relative_path=declarative_artifact.relative_path
+                            )
+                            content_artifact_bulk.append(content_artifact)
+                            remote_artifact_data = {
+                                'url': declarative_artifact.url,
+                                'size': declarative_artifact.artifact.size,
+                                'md5': declarative_artifact.artifact.md5,
+                                'sha1': declarative_artifact.artifact.sha1,
+                                'sha224': declarative_artifact.artifact.sha224,
+                                'sha256': declarative_artifact.artifact.sha256,
+                                'sha384': declarative_artifact.artifact.sha384,
+                                'sha512': declarative_artifact.artifact.sha512,
+                                'remote': declarative_artifact.remote,
+                            }
+                            rel_path = content_artifact.relative_path
+                            remote_artifact_map[rel_path] = remote_artifact_data
```

Comment:
The `remote_artifact_map` *dict* seems to be shared across `Content` but *keyed* by artifact relative path.  What if two different artifacts in two different content has the same relative path?

Reply:
I see what you're saying for Pulp generally, but for the pipelines, I don't think that is going to happen because it's an invalid repo. For that to happen in the stream, it would effectively be wanting to include two Artifacts with the same relative path in one repo which would fail in various places anyway.

Reply:
Perhaps but the result here is silently creating a remote artifact with the (wrong) information from another (remote) artifact.  That seems unacceptable.

Reply:
Wait I'm confused. It's not happening because in practice you can't create a stream that creates two Artifacts at one relative_path. So there won't be duplicates like that.

Let me ask you this, what do you think should be done instead that is safer?

Reply:
How creating two artifacts (on different content) prevented?